### ISSUE TITLE
linter: React compiler invariant violation when using computed property destructuring

### ISSUE THREAD

**Issue by TheNumberOne** [oxc-project/oxc#26090]
### What version of Oxlint are you using?

1.79.0

### What command did you run?

`oxlint`

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```json
{
  "plugins": ["react"],
  "rules": {
    "react/invariant": "error"
  }
}
```

### What happened?

When I run the `react/invariant` lint on code that includes computed property destructuring, I get a failure from the lint: `react(invariant): [InferMutationAliasingEffects] Expected value kind to be initialized`. It seems like the purpose of this lint is to indicate bugs in oxc's port of the react compiler, so here you go lol.

```ts
import { useMemo, useState } from 'react';

const Key = { Compact: 'compact' } as const;

export const useExample = () => {
  const [settings] = useState<Record<string, string>>({});
  const { [Key.Compact]: value } = settings;

  return useMemo(() => value ?? '', [value]);
};
```

When I run the code against the official react compiler, I get the following error. I think that means the expected behavior should be for this to trigger `react/todo` and not `react/invariant`.
```json
{
  "kind": "CompileError",
  "reason": "(BuildHIR::lowerAssignment) Handle computed properties in ObjectPattern"
}
```

Related issues: https://github.com/oxc-project/oxc/issues/23750 https://github.com/oxc-project/oxc/pull/25513

**Comment by Boshen** [comment]
Investigated this against react/react main (3c397fe76015d6988cdcb6f74aff978a3cc6a37a).

This is not currently an Oxc-specific divergence. React already has a computed object-pattern fixture that expects the same InferMutationAliasingEffects invariant, and both its TypeScript and Rust snapshot runners pass with that expectation:

- https://github.com/react/react/blob/3c397fe76015d6988cdcb6f74aff978a3cc6a37a/compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/todo.error.object-pattern-computed-key.expect.md#L21-L34
- https://github.com/oxc-project/oxc/blob/de31accb7a143fb6faafcf612c0991aeca11abe0/crates/oxc_react_compiler/tests/snapshots/todo.error.object-pattern-computed-key.snap#L5-L9

The failure path is:

1. BuildHIR records the Todo for the unsupported computed property and skips that property.
2. Lowering continues with partial HIR, leaving value uninitialized.
3. InferMutationAliasingEffects later reads value and throws the invariant, which supersedes the accumulated Todo.

The Todo result from the older official compiler matches React behavior before react/react#35878. That fault-tolerance change made lower() always produce partial HIR and changed the upstream snapshot from Todo to this invariant.

So oxlint 1.79.0 is matching current upstream behavior. I suggest closing this as upstream parity, or reframing it as an upstream React diagnostic regression. Changing only Oxc to report react/todo would intentionally diverge; the proper fix would be to make the partial HIR valid or prevent the later invariant upstream.

AI-assisted investigation; I reviewed the relevant source, history, reproductions, and snapshot results.

**Comment by TheNumberOne** [comment]
ah, i didn't realize the upstream error was an invariant one! I thought it was a todo. makes sense
